File: bindings/python/satox_bindings/ipfs_manager.py
"""
Satox SDK IPFS Manager Binding
"""

from typing import Dict, List, Optional, Any
import asyncio
from .error import SatoxError

class IPFSManager:
    """Wrapper for the C++ IPFSManager"""
    
    def __init__(self):
        self._core = None
        self._initialized = False
        self._callbacks = []
    
    async def initialize(self, config: Dict[str, Any]) -> bool:
        """Initialize IPFS manager"""
        if self._initialized:
            raise SatoxError("IPFSManager already initialized")
        
        try:
            result = await self._call_native("initialize", config)
            self._initialized = result
            return result
        except Exception as e:
            raise SatoxError(f"Failed to initialize IPFSManager: {str(e)}")
    
    async def upload_file(self, file_path: str) -> str:
        """Upload file to IPFS"""
        self._ensure_initialized()
        
        try:
            return await self._call_native("upload_file", file_path)
        except Exception as e:
            raise SatoxError(f"Failed to upload file: {str(e)}")
    
    async def download_file(self, ipfs_hash: str, output_path: str) -> bool:
        """Download file from IPFS"""
        self._ensure_initialized()
        
        try:
            return await self._call_native("download_file", ipfs_hash, output_path)
        except Exception as e:
            raise SatoxError(f"Failed to download file: {str(e)}")
    
    async def upload_data(self, data: bytes) -> str:
        """Upload data to IPFS"""
        self._ensure_initialized()
        
        try:
            return await self._call_native("upload_data", data)
        except Exception as e:
            raise SatoxError(f"Failed to upload data: {str(e)}")
    
    async def get_data(self, ipfs_hash: str) -> bytes:
        """Get data from IPFS"""
        self._ensure_initialized()
        
        try:
            return await self._call_native("get_data", ipfs_hash)
        except Exception as e:
            raise SatoxError(f"Failed to get data: {str(e)}")
    
    def _ensure_initialized(self):
        if not self._initialized:
            raise SatoxError("IPFSManager not initialized")
    
    async def _call_native(self, method: str, *args) -> Any:
        """Call native C++ method"""
        raise NotImplementedError("Native binding not implemented") 

File: bindings/typescript/src/ipfs-manager.ts
/**
 * Satox SDK IPFS Manager Binding
 */

import { SatoxError } from './error';

export class IPFSManager {
    private core: any;
    private initialized: boolean = false;
    private callbacks: Array<(event: string, data: any) => void> = [];

    constructor() {
        this.core = require('./native/satox_core');
    }

    async initialize(config: Record<string, any>): Promise<boolean> {
        if (this.initialized) {
            throw new SatoxError('IPFSManager already initialized');
        }

        try {
            const result = await this.core.ipfsInitialize(config);
            this.initialized = result;
            return result;
        } catch (error) {
            throw new SatoxError(`Failed to initialize IPFSManager: ${error.message}`);
        }
    }

    async uploadFile(filePath: string): Promise<string> {
        this.ensureInitialized();

        try {
            return await this.core.ipfsUploadFile(filePath);
        } catch (error) {
            throw new SatoxError(`Failed to upload file: ${error.message}`);
        }
    }

    async downloadFile(ipfsHash: string, outputPath: string): Promise<boolean> {
        this.ensureInitialized();

        try {
            return await this.core.ipfsDownloadFile(ipfsHash, outputPath);
        } catch (error) {
            throw new SatoxError(`Failed to download file: ${error.message}`);
        }
    }

    async uploadData(data: Buffer): Promise<string> {
        this.ensureInitialized();

        try {
            return await this.core.ipfsUploadData(data);
        } catch (error) {
            throw new SatoxError(`Failed to upload data: ${error.message}`);
        }
    }

    async getData(ipfsHash: string): Promise<Buffer> {
        this.ensureInitialized();

        try {
            return await this.core.ipfsGetData(ipfsHash);
        } catch (error) {
            throw new SatoxError(`Failed to get data: ${error.message}`);
        }
    }

    registerCallback(callback: (event: string, data: any) => void): void {
        this.callbacks.push(callback);
    }

    unregisterCallback(callback: (event: string, data: any) => void): void {
        const index = this.callbacks.indexOf(callback);
        if (index > -1) {
            this.callbacks.splice(index, 1);
        }
    }

    private ensureInitialized(): void {
        if (!this.initialized) {
            throw new SatoxError('IPFSManager not initialized');
        }
    }
} 